File: src/ParserNbBet/Scheduler/MskScheduler.cs
using ParserNbBet.Config;
using Serilog;

namespace ParserNbBet.Scheduler;

/// <summary>
/// Computes next run times based on MSK start time + interval,
/// and runs a loop using PeriodicTimer with graceful shutdown.
/// </summary>
public sealed class MskScheduler : IDisposable
{
    private static readonly TimeZoneInfo MskZone =
        TimeZoneInfo.FindSystemTimeZoneById("Russian Standard Time");

    private readonly ScheduleConfig _config;
    private readonly Func<CancellationToken, Task> _cycleAction;
    private CancellationTokenSource? _cts;

    /// <summary>Fires when the next-run time is computed (for UI).</summary>
    public event Action<DateTime>? NextRunComputed;

    /// <summary>Fires when a cycle starts.</summary>
    public event Action? CycleStarted;

    /// <summary>Fires when a cycle finishes (bool = success).</summary>
    public event Action<bool>? CycleFinished;

    public MskScheduler(ScheduleConfig config, Func<CancellationToken, Task> cycleAction)
    {
        _config = config ?? throw new ArgumentNullException(nameof(config));
        _cycleAction = cycleAction ?? throw new ArgumentNullException(nameof(cycleAction));
    }

    /// <summary>
    /// Runs a single cycle (for --once mode).
    /// </summary>
    public async Task RunOnceAsync(CancellationToken ct = default)
    {
        CycleStarted?.Invoke();
        bool ok = false;
        try
        {
            await _cycleAction(ct);
            ok = true;
        }
        finally
        {
            CycleFinished?.Invoke(ok);
        }
    }

    /// <summary>
    /// Runs the daemon loop: compute next slot, wait, execute cycle, repeat.
    /// </summary>
    public async Task RunDaemonAsync(CancellationToken ct = default)
    {
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        var token = _cts.Token;

        Log.Information("Daemon started. Schedule: start={Start} MSK, interval={Interval}h",
            _config.StartTimeMsk, _config.IntervalHours);

        while (!token.IsCancellationRequested)
        {
            var nowUtc = DateTime.UtcNow;
            var nextUtc = ComputeNextRunUtc(nowUtc, _config);

            Log.Information("Next run at {NextMsk} MSK ({NextUtc} UTC)",
                TimeZoneInfo.ConvertTimeFromUtc(nextUtc, MskZone).ToString("yyyy-MM-dd HH:mm:ss"),
                nextUtc.ToString("yyyy-MM-dd HH:mm:ss"));

            NextRunComputed?.Invoke(nextUtc);

            var delay = nextUtc - nowUtc;
            if (delay > TimeSpan.Zero)
            {
                try
                {
                    await Task.Delay(delay, token);
                }
                catch (OperationCanceledException)
                {
                    break;
                }
            }

            if (token.IsCancellationRequested) break;

            // Execute cycle
            CycleStarted?.Invoke();
            bool ok = false;
            try
            {
                Log.Information("Cycle starting...");
                await _cycleAction(token);
                ok = true;
                Log.Information("Cycle complete.");
            }
            catch (OperationCanceledException)
            {
                Log.Information("Cycle cancelled.");
                break;
            }
            catch (Exception ex)
            {
                Log.Error(ex, "Cycle failed");
            }
            finally
            {
                CycleFinished?.Invoke(ok);
            }
        }

        Log.Information("Daemon stopped.");
    }

    /// <summary>
    /// Request graceful shutdown.
    /// </summary>
    public void Stop()
    {
        _cts?.Cancel();
    }

    /// <summary>
    /// Computes the next run time (UTC) based on MSK schedule config.
    /// Schedule slots are: startTimeMsk, startTimeMsk + interval, startTimeMsk + 2*interval, ...
    /// Returns the nearest future slot.
    /// </summary>
    public static DateTime ComputeNextRunUtc(DateTime nowUtc, ScheduleConfig config)
    {
        var nowMsk = TimeZoneInfo.ConvertTimeFromUtc(nowUtc, MskZone);

        // Parse start time
        var parts = config.StartTimeMsk.Split(':');
        int startHour = int.Parse(parts[0]);
        int startMinute = parts.Length > 1 ? int.Parse(parts[1]) : 0;

        var intervalHours = config.IntervalHours;

        // Generate all slots for today and tomorrow, find the next one
        for (int dayOffset = 0; dayOffset <= 1; dayOffset++)
        {
            var day = nowMsk.Date.AddDays(dayOffset);
            var slot = new DateTime(day.Year, day.Month, day.Day, startHour, startMinute, 0);

            while (slot.Hour < 24)
            {
                if (slot > nowMsk)
                {
                    // Convert back to UTC
                    return TimeZoneInfo.ConvertTimeToUtc(
                        DateTime.SpecifyKind(slot, DateTimeKind.Unspecified), MskZone);
                }

                slot = slot.AddHours(intervalHours);
                // If we go past midnight, break to next day
                if (slot.Date > day.Date)
                    break;
            }
        }

        // Fallback: first slot tomorrow (should not happen with dayOffset=1 above)
        var tomorrow = nowMsk.Date.AddDays(1);
        var fallback = new DateTime(tomorrow.Year, tomorrow.Month, tomorrow.Day, startHour, startMinute, 0);
        return TimeZoneInfo.ConvertTimeToUtc(
            DateTime.SpecifyKind(fallback, DateTimeKind.Unspecified), MskZone);
    }

    public void Dispose()
    {
        _cts?.Cancel();
        _cts?.Dispose();
    }
}

File: src/scheduler/msk_scheduler.py
"""MSK timezone scheduler.

Computes next run times based on MSK (Europe/Moscow) start time + interval.
Supports --once and --daemon modes with graceful shutdown.
"""
from __future__ import annotations

import logging
import threading
import time
from datetime import datetime, timedelta, timezone

log = logging.getLogger("parser_nb_bet.scheduler.msk_scheduler")

# MSK = UTC+3 (fixed, no DST since 2014)
MSK = timezone(timedelta(hours=3))


def now_msk() -> datetime:
    """Current time in MSK."""
    return datetime.now(MSK)


def parse_start_time(time_str: str) -> tuple[int, int]:
    """Parse 'HH:MM' start time string. Returns (hour, minute)."""
    parts = time_str.split(":")
    if len(parts) != 2:
        raise ValueError(f"Invalid start_time format: {time_str!r}, expected HH:MM")
    return int(parts[0]), int(parts[1])


def compute_next_run(
    start_hour: int,
    start_minute: int,
    interval_hours: int,
    ref_time: datetime | None = None,
) -> datetime:
    """Compute the next scheduled run time in MSK.

    Generates all slots for today: start_time, start_time + interval, ...
    Returns the first slot that is in the future relative to ref_time.
    If all today's slots have passed, returns first slot tomorrow.
    """
    if ref_time is None:
        ref_time = now_msk()
    else:
        ref_time = ref_time.astimezone(MSK)

    today = ref_time.date()

    # Generate slots for today
    base = datetime(today.year, today.month, today.day, start_hour, start_minute, tzinfo=MSK)
    slots = []
    slot = base
    while slot.date() == today:
        slots.append(slot)
        slot = slot + timedelta(hours=interval_hours)

    # Find next future slot
    for s in slots:
        if s > ref_time:
            return s

    # All passed — first slot tomorrow
    tomorrow_base = base + timedelta(days=1)
    return tomorrow_base


class MskScheduler:
    """Scheduler that runs cycles at MSK-based time slots.

    Args:
        start_time_msk: Start time string 'HH:MM'.
        interval_hours: Hours between runs.
        shutdown_event: Threading event for graceful shutdown.
    """

    def __init__(
        self,
        start_time_msk: str = "08:00",
        interval_hours: int = 4,
        shutdown_event: threading.Event | None = None,
    ) -> None:
        self.start_hour, self.start_minute = parse_start_time(start_time_msk)
        self.interval_hours = interval_hours
        self.shutdown_event = shutdown_event or threading.Event()

    def next_run(self, ref_time: datetime | None = None) -> datetime:
        """Get next scheduled run time."""
        return compute_next_run(
            self.start_hour, self.start_minute, self.interval_hours, ref_time
        )

    def wait_until_next(self, ref_time: datetime | None = None) -> bool:
        """Wait until next scheduled run. Returns False if shutdown requested."""
        target = self.next_run(ref_time)
        now = now_msk()
        wait_secs = (target - now).total_seconds()

        if wait_secs <= 0:
            return True

        log.info(
            "Next run at %s MSK (in %.0f min)",
            target.strftime("%H:%M:%S"),
            wait_secs / 60,
        )

        # Wait with periodic checks for shutdown
        return not self.shutdown_event.wait(timeout=wait_secs)

    def run_daemon(self, cycle_fn: callable) -> None:
        """Run cycle_fn on schedule until shutdown.

        Args:
            cycle_fn: Callable that executes one cycle. Called with no args.
        """
        log.info(
            "Daemon started: MSK %02d:%02d every %dh",
            self.start_hour, self.start_minute, self.interval_hours,
        )

        while not self.shutdown_event.is_set():
            if not self.wait_until_next():
                log.info("Shutdown requested, exiting daemon")
                break

            if self.shutdown_event.is_set():
                break

            log.info("Cycle starting at %s MSK", now_msk().strftime("%H:%M:%S"))
            try:
                cycle_fn()
            except Exception:
                log.exception("Cycle failed")

            # Small delay to avoid re-triggering same slot
            time.sleep(2)

        log.info("Daemon stopped")

    def request_shutdown(self) -> None:
        """Request graceful shutdown."""
        self.shutdown_event.set()
